// wiki-plugin-grep/client/grep.js

/*
 * Federated Wiki : Grep Plugin
 *
 * Licensed under the MIT license.
 * https://github.com/fedwiki/wiki-plugin-grep/blob/master/LICENSE.txt
 */

(function() {
  var bind, emit, escape, evalPage, evalPart, parse, run, word;

  escape = function(line) {
    return line.replace(/&/g, '&amp;').replace(/</g, '&lt;').replace(/>/g, '&gt;');
  };

  word = function(string) {
    if (!string.match(/^[a-z]*$/)) {
      throw {
        message: "expecting type for '" + string + "'"
      };
    }
    return string;
  };

  parse = function(text) {
    var arg, err, error, errors, html, i, len, line, listing, match, op, program, ref, ref1;
    program = [];
    listing = [];
    errors = 0;
    ref = text.split("\n");
    for (i = 0, len = ref.length; i < len; i++) {
      line = ref[i];
      html = escape(line);
      try {
        ref1 = line.match(/^\s*(\w*)\s*(.*)$/), match = ref1[0], op = ref1[1], arg = ref1[2];
        switch (op) {
          case '':
            break;
          case 'ITEM':
          case 'ACTION':
            program.push({
              op: op,
              type: word(arg)
            });
            break;
          case 'TEXT':
          case 'TITLE':
          case 'SITE':
          case 'ID':
          case 'ALIAS':
          case 'JSON':
            program.push({
              op: op,
              regex: new RegExp(arg, 'mi')
            });
            break;
          default:
            throw {
              message: "don't know '" + op + "' command"
            };
        }
      } catch (error) {
        err = error;
        errors++;
        html = "<span style=\"background-color:#fdd;width:100%;\" title=\"" + err.message + "\">" + html + "</span>";
      }
      listing.push(html);
    }
    return [program, listing.join('<br>'), errors];
  };

  evalPage = function(page, steps, count) {
    var action, i, item, j, len, len1, ref, ref1, step;
    if (!(count < steps.length)) {
      return true;
    }
    step = steps[count];
    switch (step.op) {
      case 'ITEM':
        count++;
        ref = page.story || [];
        for (i = 0, len = ref.length; i < len; i++) {
          item = ref[i];
          if (step.type === '') {
            if (evalPart(item, steps, count)) {
              return true;
            }
          } else {
            if (item.type === step.type) {
              if (evalPart(item, steps, count)) {
                return true;
              }
            }
          }
        }
        return false;
      case 'ACTION':
        count++;
        ref1 = page.journal || [];
        for (j = 0, len1 = ref1.length; j < len1; j++) {
          action = ref1[j];
          if (step.type === '') {
            if (evalPart(action, steps, count)) {
              return true;
            }
          } else {
            if (action.type === step.type) {
              if (evalPart(action, steps, count)) {
                return true;
              }
            }
          }
        }
        return false;
    }
    return evalPart(page, steps, count);
  };

  evalPart = function(part, steps, count) {
    var json, key, ref, step;
    if (!(count < steps.length)) {
      return true;
    }
    step = steps[count++];
    switch (step.op) {
      case 'TEXT':
      case 'TITLE':
      case 'SITE':
      case 'ID':
      case 'ALIAS':
        key = step.op.toLowerCase();
        if ((part[key] || ((ref = part.item) != null ? ref[key] : void 0) || '').match(step.regex)) {
          return true;
        }
        break;
      case 'JSON':
        json = JSON.stringify(part, null, ' ');
        if (json.match(step.regex)) {
          return true;
        }
    }
    return false;
  };

  run = function($item, program) {
    var status, want;
    status = function(text) {
      return $item.find('.caption').text(text);
    };
    want = function(page) {
      return evalPage(page, program, 0);
    };
    status("fetching sitemap");
    return $.getJSON("http://" + location.host + "/system/sitemap.json", function(sitemap) {
      var checked, found, i, len, place, results;
      checked = 0;
      found = 0;
      results = [];
      for (i = 0, len = sitemap.length; i < len; i++) {
        place = sitemap[i];
        results.push($.getJSON("http://" + location.host + "/" + place.slug + ".json", function(page) {
          var report, text;
          text = "[[" + page.title + "]] (" + page.story.length + ")";
          if (want(page)) {
            found++;
            $item.find('.result').append((wiki.resolveLinks(text)) + "<br>");
          }
          checked++;
          report = "found " + found + " pages of " + checked + " checked";
          if (checked < sitemap.length) {
            report += ", " + (sitemap.length - checked) + " remain";
          }
          return status(report);
        }));
      }
      return results;
    });
  };

  emit = function($item, item) {
    var errors, listing, program, ref;
    ref = parse(item.text), program = ref[0], listing = ref[1], errors = ref[2];
    $item.append("<div style=\"background-color:#eee;padding:15px;\">\n  <div class=listing>" + listing + "</div>\n  <p class=\"caption\">" + errors + " errors</p>\n  <p class=\"result\"></p>\n</div>");
    if (!errors) {
      return run($item, program);
    }
  };

  bind = function($item, item) {
    return $item.dblclick(function() {
      return wiki.textEditor($item, item);
    });
  };

  if (typeof window !== "undefined" && window !== null) {
    window.plugins.grep = {
      emit: emit,
      bind: bind
    };
  }

  if (typeof module !== "undefined" && module !== null) {
    module.exports = {
      parse: parse,
      evalPart: evalPart,
      evalPage: evalPage
    };
  }

}).call(this);
